File: 1143-longest-common-subsequence/1143-longest-common-subsequence.java
class Solution {
    int[][] dp;
    public int longestCommonSubsequence(String text1, String text2) {
        dp = new int[text1.length()][text2.length()];
        for(int i=0;i<dp.length;i++){
            for(int j=0;j<dp[i].length;j++){
                dp[i][j]=-1;
            }
        }
        return helper(text1,text2,0,0);
    }
    public int helper(String s1,String s2,int i,int j){
        if(i==s1.length() || j==s2.length()){
            return 0;
        }
        if(dp[i][j]>-1){
            return dp[i][j];
        }
        if(s1.charAt(i)==s2.charAt(j)){
            return 1+helper(s1,s2,i+1,j+1);
        }
        dp[i][j] = Math.max(helper(s1,s2,i+1,j),helper(s1,s2,i,j+1));
        return dp[i][j];
    }
}

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.cpp
class Solution {
public:
    int bottomup(string text1,string text2,int m,int n,vector<vector<int>>dp){
        for(int i=m-1;i>=0;i--){
            for(int j=n-1;j>=0;j--){
                if(text1[i]==text2[j]){
                    dp[i][j]=1+dp[i+1][j+1];
                }
                else{
                    dp[i][j]=max(dp[i+1][j],dp[i][j+1]);
                }
            }
        }
        return dp[0][0];
    }
    int longestCommonSubsequence(string text1, string text2) {
        int m=text1.size();
        int n=text2.size();
        vector<vector<int>> dp(m+1,vector<int>(n+1,0));
        return bottomup(text1,text2,m,n,dp);
    }
};